### PR TITLE
Hash set

### PR DESCRIPTION
**Pull request by dmurasov**
My personal implementation of HashSet.

### CODE REVIEW

File: src/main/java/com/endava/internship/collections/StudentSet.java [hunk 1]
```diff
@@ -1,84 +1,328 @@
...
-        //Ignore this for homework
-        throw new UnsupportedOperationException();
+        //throwing exception if null
+        Objects.requireNonNull(collection);
+        boolean removed = false;
+        Iterator<?> stIterator = iterator();
+        while (stIterator.hasNext()){
+            Student st = (Student)stIterator.next();
+            if (collection.contains(st)){
+                System.out.println(st);
```
Comment (perepiolcamaxim) [inline src/main/java/com/endava/internship/collections/StudentSet.java]: please remove println

File: src/main/java/com/endava/internship/collections/StudentSet.java [hunk 2]
```diff
@@ -1,84 +1,328 @@
...
+    private void setCapacity(int capacity){
+        this.capacity = capacity;
+    }
+    public StudentSet(int capacity){
+        this.buckets = new Entry[capacity];
+        setCapacity(capacity);
+    }
+    public StudentSet(){
+        this(INITIAL_CAPACITY);
+    }
```
Comment (perepiolcamaxim) [inline src/main/java/com/endava/internship/collections/StudentSet.java]: please move constructors after instance fields

File: src/main/java/com/endava/internship/collections/StudentSet.java [hunk 3]
```diff
@@ -1,84 +1,328 @@
...
-import java.util.Collection;
-import java.util.Iterator;
-import java.util.Set;
+import java.util.*;
+import java.lang.reflect.Array;
 
 public class StudentSet implements Set<Student> {
+    private Entry<Student>[] buckets;
+    private int capacity, size;
+    private static final int INITIAL_CAPACITY = 16;
```
Comment (perepiolcamaxim) [inline src/main/java/com/endava/internship/collections/StudentSet.java]: static fields first

Comment (perepiolcamaxim) [reply src/main/java/com/endava/internship/collections/StudentSet.java]: and also add a new line after each statement

File: src/main/java/com/endava/internship/collections/StudentSet.java [hunk 4]
```diff
@@ -1,84 +1,328 @@
...
     public boolean isEmpty() {
-        //TODO
-        return false;
+        return size() == 0;
     }
 
     @Override
     public boolean contains(Object o) {
-        //TODO
+        int i = o == null ? 0 : hash(o.hashCode());
```
Comment (perepiolcamaxim) [inline src/main/java/com/endava/internship/collections/StudentSet.java]: move this case to hash function

Comment (perepiolcamaxim) [reply src/main/java/com/endava/internship/collections/StudentSet.java]: make hash to take a Student param

**Review by perepiolcamaxim — APPROVED** [review]

File: src/main/java/com/endava/internship/collections/StudentSet.java (hunk 1 above)
Comment (dmurasov) [reply src/main/java/com/endava/internship/collections/StudentSet.java]: Fixed

File: src/main/java/com/endava/internship/collections/StudentSet.java (hunk 2 above)
Comment (dmurasov) [reply src/main/java/com/endava/internship/collections/StudentSet.java]: Fixed

File: src/main/java/com/endava/internship/collections/StudentSet.java (hunk 3 above)
Comment (dmurasov) [reply src/main/java/com/endava/internship/collections/StudentSet.java]: Fixed

File: src/main/java/com/endava/internship/collections/StudentSet.java (hunk 4 above)
Comment (dmurasov) [reply src/main/java/com/endava/internship/collections/StudentSet.java]: Fixed

**Pull request merged**